sgkit/window.py

```diff
@@ -0,0 +1,174 @@
+
+
+def moving_statistic(
+    values: Any,
+    statistic: Callable[..., Any],
+    size: int,
+    step: int,
+    dtype: int,
+    **kwargs: Any,
+) -> Any:
+    """A Dask implementation of scikit-allel's moving_statistic function."""
+    length = values.shape[0]
+    chunks = values.chunks[0]
+    if len(chunks) > 1:
+        min_chunksize = np.min(chunks[:-1])  # ignore last chunk
+    else:
+        min_chunksize = np.min(chunks)
+    if min_chunksize < size:
+        raise ValueError(
+            f"Minimum chunk size ({min_chunksize}) must not be smaller than size ({size})."
+        )
+    window_starts, window_stops = _get_windows(0, length, size, step)
+    return window_statistic(
+        values, statistic, window_starts, window_stops, dtype, **kwargs
+    )
+
+
+def window_statistic(
+    values: Any,
+    statistic: Callable[..., Any],
+    window_starts: Any,
+    window_stops: Any,
+    dtype: int,
+    **kwargs: Any,
+) -> Any:
+
+    values = da.asarray(values)
+
+    length = values.shape[0]
+    window_lengths = window_stops - window_starts
+    depth = np.max(window_lengths)
+
+    # Dask will raise an error if the last chunk size is smaller than the depth
+    # Workaround by rechunking to combine the last two chunks in first axis
+    # See https://github.com/dask/dask/issues/6597
+    if depth > values.chunks[0][-1]:
+        chunk0 = values.chunks[0]
+        new_chunk0 = tuple(list(chunk0[:-2]) + [chunk0[-2] + chunk0[-1]])
+        # None means don't rechunk along that axis
+        new_chunks = tuple(list([new_chunk0] + ([None] * (len(chunk0) - 1))))  # type: ignore
+        values = values.rechunk(new_chunks)
+
+    chunks = values.chunks[0]
+
+    rel_window_starts, windows_per_chunk = _get_chunked_windows(
+        chunks, length, window_starts, window_stops
+    )
+
+    # Add depth for map_overlap
+    rel_window_starts = rel_window_starts + depth
+    rel_window_stops = rel_window_starts + window_lengths
+
+    chunk_offsets = _sizes_to_start_offsets(windows_per_chunk)
+
+    def blockwise_moving_stat(x: Any, block_info: Any = None) -> Any:
+        if block_info is None or len(block_info) == 0:
+            return np.array([])
+        chunk_number = block_info[0]["chunk-location"][0]
+        chunk_offset_start = chunk_offsets[chunk_number]
+        chunk_offset_stop = chunk_offsets[chunk_number + 1]
+        chunk_window_starts = rel_window_starts[chunk_offset_start:chunk_offset_stop]
+        chunk_window_stops = rel_window_stops[chunk_offset_start:chunk_offset_stop]
+        out = np.array(
+            [
+                statistic(x[i:j], **kwargs)
+                for i, j in zip(chunk_window_starts, chunk_window_stops)
+            ]
+        )
+        return out
+
+    if values.ndim == 1:
+        new_chunks = (tuple(windows_per_chunk),)
+    else:
+        # depth is 0 except in first axis
+        depth = tuple([depth] + ([0] * (values.ndim - 1)))
```

Comment:
I wonder if `map_overlap` would work in this case with `depth = {0: depth}` (without the need to specify the depth on the other dimensions) 🤔

Reply:
That should work: https://github.com/dask/dask/blob/master/dask/array/overlap.py#L732 is the relevant bit for that (all others default to 0).

Reply:
Thanks for checking - fixed